pkg/selector/emr.go

```diff
@@ -0,0 +1,342 @@
+	instanceTypes, err := e.getEMRInstanceTypes(version)
+	if err != nil {
+		return filters, err
+	}
+	filters.InstanceTypes = &instanceTypes
+	filters.RootDeviceType = aws.String("ebs")
+	filters.VirtualizationType = aws.String("hvm")
+	return filters, nil
+}
+
+// getEMRInstanceTypes returns a list of instance types that emr supports
+func (e EMR) getEMRInstanceTypes(version string) ([]string, error) {
+	instanceTypes := []string{}
+	if version == "" {
+		version = "5.20.0"
+	}
+	versionInt, err := strconv.Atoi(strings.ReplaceAll(version, ".", ""))
+	if err != nil {
+		return instanceTypes, err
+	}
+
+	for _, instanceType := range e.getAllEMRInstanceTypes() {
+		if versionInt >= 5250 {
+			instanceTypes = append(instanceTypes, instanceType)
+		} else if versionInt >= 5200 && versionInt < 5250 {
+			if e.isOnlyEMR_5_25_0_plus(instanceType) {
+				continue
+			}
+			instanceTypes = append(instanceTypes, instanceType)
+		} else if versionInt >= 5150 && versionInt < 5200 {
+			if instanceType == "c1.medium" {
+				continue
+			}
+			if e.isOnlyEMR_5_20_0_plus(instanceType) {
+				continue
+			}
+			if e.isOnlyEMR_5_25_0_plus(instanceType) {
+				continue
+			}
+			instanceTypes = append(instanceTypes, instanceType)
+		} else if versionInt >= 5130 && versionInt < 5150 {
+			if e.isOnlyEMR_5_20_0_plus(instanceType) {
+				continue
+			}
+			if e.isOnlyEMR_5_25_0_plus(instanceType) {
+				continue
+			}
+			instanceTypes = append(instanceTypes, instanceType)
+		} else if versionInt >= 590 && versionInt < 5130 {
+			if e.isEMR_5_13_0_plus(instanceType) {
+				continue
+			}
+			if e.isOnlyEMR_5_20_0_plus(instanceType) {
+				continue
+			}
+			if e.isOnlyEMR_5_25_0_plus(instanceType) {
+				continue
+			}
+			instanceTypes = append(instanceTypes, instanceType)
+		} else {
+			if e.isEMR_5_13_0_plus(instanceType) {
+				continue
+			}
+			if e.isOnlyEMR_5_20_0_plus(instanceType) {
+				continue
+			}
+			if e.isOnlyEMR_5_25_0_plus(instanceType) {
+				continue
+			}
+			if strings.HasPrefix(instanceType, "i3") {
+				continue
+			}
+			instanceTypes = append(instanceTypes, instanceType)
+		}
+	}
+	return instanceTypes, nil
+}
+
+func (EMR) isEMR_5_13_0_plus(instanceType string) bool {
+	prefixes := []string{
+		"m5.",
+		"m5d.",
+		"c5.",
+		"c5d.",
+		"r5.",
+		"r5d.",
+	}
+	for _, prefix := range prefixes {
+		if strings.HasPrefix(instanceType, prefix) {
+			return true
+		}
+	}
+	return false
+}
+
+func (EMR) isOnlyEMR_5_20_0_plus(instanceType string) bool {
+	prefixes := []string{
+		"m5a.",
+		"c5n.",
+		"r5a.",
+	}
+	for _, prefix := range prefixes {
+		if strings.HasPrefix(instanceType, prefix) {
+			return true
+		}
+	}
+	return false
+}
+
+func (EMR) isOnlyEMR_5_25_0_plus(instanceType string) bool {
+	prefixes := []string{
+		"i3en.",
+	}
+	for _, prefix := range prefixes {
+		if strings.HasPrefix(instanceType, prefix) {
+			return true
+		}
+	}
+	return false
+}
+
+func (EMR) getAllEMRInstanceTypes() []string {
+	return []string{
```

Comment:
how often will this need to be updated?

Reply:
I'm not sure.... This is definitely not an optimal solution, but I think it'll be the best hack available. Other downsides are that users will have to actually update the ec2-instance-selector binary to get updated versions or to add supported instance types to older versions.

Reply:
Gotcha, if we see ourselves needing to manually update this list often then we could look into watching and/or pulling directly from the [github readme on emr types](https://github.com/awsdocs/amazon-emr-management-guide/blob/master/doc_source/emr-supported-instance-types.md)